datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+/// ```
+///
+/// which is displayed as follows in datafusion-cli:
+///
+/// ```text
+/// 2019-03-31T01:00:00
+/// ```
+fn adjust_to_local_time<T: ArrowTimestampType>(ts: i64, timezone: &str) -> i64 {
+    let date_time = match T::UNIT {
+        Nanosecond => Utc.timestamp_nanos(ts),
+        Microsecond => Utc.timestamp_micros(ts).unwrap(), // TODO chunchun: replace unwrap
+        Millisecond => Utc.timestamp_millis_opt(ts).unwrap(),
+        Second => Utc.timestamp_opt(ts, 0).unwrap(),
+    };
+
+    let tz: Tz = timezone.parse().unwrap();
+    let offset_seconds: i64 = tz
+        .offset_from_utc_datetime(&date_time.naive_utc())
+        .fix()
+        .local_minus_utc() as i64;
+
+    let adjusted_date_time =
+        date_time.add(TimeDelta::try_seconds(offset_seconds).unwrap());
+
+    // convert the navie datetime back to i64
+    match T::UNIT {
+        Nanosecond => adjusted_date_time.timestamp_nanos_opt().unwrap(), // TODO chuynchun: remove unwrap
+        Microsecond => adjusted_date_time.timestamp_micros(),
+        Millisecond => adjusted_date_time.timestamp_millis(),
+        Second => adjusted_date_time.timestamp(),
+    }
+}
+
+impl ScalarUDFImpl for ToLocalTimeFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "to_local_time"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if arg_types.len() != 1 {
+            return exec_err!(
+                "to_local_time function requires 1 argument, got {:?}",
+                arg_types.len()
+            );
+        }
+
+        match &arg_types[0] {
+            Timestamp(Nanosecond, _) => Ok(Timestamp(Nanosecond, None)),
+            Timestamp(Microsecond, _) => Ok(Timestamp(Microsecond, None)),
+            Timestamp(Millisecond, _) => Ok(Timestamp(Millisecond, None)),
+            Timestamp(Second, _) => Ok(Timestamp(Second, None)),
+            _ => exec_err!(
+                "The to_local_time function can only accept timestamp as the arg, got {:?}", arg_types[0]
+            ),
+        }
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        if args.len() != 1 {
+            return exec_err!(
+                "to_local_time function requires 1 argument, got {:?}",
+                args.len()
+            );
+        }
+
+        self.to_local_time(args)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use arrow::array::{types::TimestampNanosecondType, TimestampNanosecondArray};
+    use arrow::compute::kernels::cast_utils::string_to_timestamp_nanos;
+    use arrow::datatypes::{DataType, TimeUnit};
+    use chrono::NaiveDateTime;
+    use datafusion_common::ScalarValue;
+    use datafusion_expr::{ColumnarValue, ScalarUDFImpl};
+
+    use super::ToLocalTimeFunc;
+
+    #[test]
+    fn test_to_local_time() {
```

Comment:
I think we can add the tests to `datafusion/sqllogictest/test_files/functions.slt`

Reply:
Yes, please!

Reply:
In general, if possible I think we should use .slt for these tests. The instructions are in https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

If there is something we can't test easily / completely with .slt tests then I think it is ok to use these unit tests, but it would be good to add a comment about why they aren't tested via .slt

For example the error tests are probably good to put in slt

Reply:
@jayzhan211 @alamb Thank you for the suggestions! I found `datafusion/sqllogictest/test_files/timestamps.slt` is more fitting to add the tests for `to_local_time()` function, so I add them there.

Updated in https://github.com/apache/datafusion/pull/11347/commits/3456517854ff0f3f533acc7e9ffcbbc7f4ee7acc